keysign/wormholeoffer.py

```diff
@@ -0,0 +1,164 @@
+from __future__ import unicode_literals
+from binascii import hexlify
+from textwrap import dedent
+from wormhole.cli.public_relay import RENDEZVOUS_RELAY
+from wormhole.errors import TransferError, WrongPasswordError
+import wormhole
+import logging
+import os
+from builtins import input
+from .gpgmh import get_usable_keys, get_public_key_data
+import gi
+gi.require_version('Gtk', '3.0')
+from gi.repository import Gtk, GLib
+if __name__ == "__main__":
+    from twisted.internet import gtk3reactor
+    gtk3reactor.install()
+from twisted.internet import reactor
+from .util import encode_message, decode_message
+
+if __name__ == "__main__" and __package__ is None:
+    logging.getLogger().error("You seem to be trying to execute " +
+                              "this script directly which is discouraged. " +
+                              "Try python -m instead.")
+    parent_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    os.sys.path.insert(0, parent_dir)
+    os.sys.path.insert(0, os.path.join(parent_dir, 'monkeysign'))
+    __package__ = str('keysign')
+
+
+log = logging.getLogger(__name__)
+
+
+class WormholeOffer:
+    def __init__(self, key, callback_receive=None, callback_code=None, app_id=None, code=None):
+        self.w = None
+        self.key = key
+        self.callback_receive = callback_receive
+        self.callback_code = callback_code
+        self.code = code
+        if app_id:
+            self.app_id = app_id
+        else:
+            # the following id is needed for interoperability with wormhole cli
+            self.app_id = "lothar.com/wormhole/text-or-file-xfer"
+
+    def start(self):
+        log.info("Wormhole: Sending a message")
+
+        self.stop()
+        self.w = wormhole.create(self.app_id, RENDEZVOUS_RELAY, reactor)
+        if self.code:
+            self.w.set_code(self.code)
+        else:
+            self.w.allocate_code()
+            self.w.get_code().addCallback(self._write_code)
+
+        # With _handle_failure we catch the WrongPasswordError
+        self.w.get_verifier().addCallbacks(self._verified, self._handle_failure)
+
+        key_data = get_public_key_data(self.key.fingerprint)
+        kd_decoded = key_data.decode('utf-8')
+        # The message needs to be encoded as a json with "message" and "offer" for ensures
+        # wormhole cli interoperability
+        offer = {"message": kd_decoded}
+        data = {"offer": offer}
+        m = encode_message(data)
+        self.w.send_message(m)
+
+        # wait for reply
+        # TODO add a timeout?
+        self.w.get_message().addCallback(self._received)
+
+    def _write_code(self, code_generated):
+        log.info("Invitation Code: {}".format(code_generated))
+        wormhole_data = "WORM={0}".format(code_generated)
+        if self.callback_code:
+            GLib.idle_add(self.callback_code, code_generated, wormhole_data)
```

Comment:
Ah, this is the full wormhole code, right?  Like not the bytes.  Using the bytes seems more appropriate to me in a machine readable code. Is it possible to encode the wormhole string code to bytes (potentially using the PGPWordlist. But maybe there is a more direct way to extract the bytes)?